File: graph_algorithms/week2_directed_graphs/3_intersection_reachability/strongly_connected.cpp
#include <iostream>
#include <stack>
#include <vector>

using std::vector;

void explore(vector<vector<int>>& adj, vector<bool>& visited, int u) {
  visited[u] = true;

  for (int v : adj[u])
    if (!visited[v])
      explore(adj, visited, v);
}

void topologicalSort(vector<vector<int>>& adj, vector<bool>& visited, std::stack<int>& s, int u) {
  visited[u] = true;

  for (int v : adj[u])
    if (!visited[v])
      topologicalSort(adj, visited, s, v);

  s.push(u);
}

int number_of_strongly_connected_components(vector<vector<int>>& adj) {
  int res = 0;

  int n = adj.size();
  vector<bool> visited(n, false);
  std::stack<int> s;

  for (int u = 0; u < n; u++)
    if (!visited[u])
      topologicalSort(adj, visited, s, u);

  vector<vector<int>> reverse_adj(n, vector<int>());
  for (int u = 0; u < n; u++)
    for (int v : adj[u])
      reverse_adj[v].push_back(u);

  visited.assign(n, false);

  while (!s.empty()) {
    int u = s.top();
    s.pop();

    if (!visited[u]) {
      explore(reverse_adj, visited, u);
      res++;
    }
  }

  return res;
}

int main() {
  int n, m;
  std::cin >> n >> m;

  vector<vector<int>> adj(n, vector<int>());
  for (int i = 0; i < m; i++) {
    int a, b;
    std::cin >> a >> b;
    adj[a - 1].push_back(b - 1);
  }

  std::cout << number_of_strongly_connected_components(adj) << std::endl;
  return 0;
}

File: graph_algorithms/week2_graph_decomposition2/3_intersection_reachability/strongly_connected.py
import sys
from typing import List

sys.setrecursionlimit(200000)


def explore(adj: List[List[int]], visited: List[bool], u: int) -> None:
    visited[u] = True
    for v in adj[u]:
        if not visited[v]:
            explore(adj, visited, v)


def topological_sort(adj: List[List[int]], visited: List[bool], stack: List[int], u: int) -> None:
    visited[u] = True
    for v in adj[u]:
        if not visited[v]:
            topological_sort(adj, visited, stack, v)
    stack.append(u)


def number_of_strongly_connected_components(adj: List[List[int]]) -> int:  # noqa: C901
    res = 0

    n = len(adj)
    visited = [False] * n
    stack: List[int] = []

    for u in range(n):
        if not visited[u]:
            topological_sort(adj, visited, stack, u)

    reverse_adj: List[List[int]] = [[] for _ in range(n)]
    for u in range(n):
        for v in adj[u]:
            reverse_adj[v].append(u)

    visited = [False] * n

    while stack:
        u = stack.pop()
        if not visited[u]:
            explore(reverse_adj, visited, u)
            res += 1

    return res


if __name__ == "__main__":
    n, m = map(int, input().split())

    edges = [tuple(map(int, input().split())) for _ in range(m)]
    adj: List[List[int]] = [[] for _ in range(n)]
    for a, b in edges:
        adj[a - 1].append(b - 1)

    print(number_of_strongly_connected_components(adj))
